### PR TITLE
Use byte arrays instead of lists of bytes in FileInfo

### LINKED ISSUE #206: FileInfo hashes use List[Byte]
`FileInfo` hashes are using `List[Byte]` to store sha1 hashes. When using `Array[Byte]`, a sha1 hash takes up approx 40 bytes of contiguous memory (including object headers etc). By contrast, a `List[Byte]` is stored as a collection of `::` nodes (but not `java.lang.Byte` instances because these are cached) and can take up to 32 bytes of memory per byte. In total, a sha1 hash stored as a `List[Byte]` can take up to 640 bytes of memory, across 20 non cached objects, requiring 41 object dereferences to iterate over. I don't know how big an impact this would actually have on performance, but it seems like such a needless waste - especially considering the classes are private and hence immutability would effectively be guaranteed when using `Array[Byte]`.

### PR DESCRIPTION
**Pull request by jroper**
Fixes #206

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=develop&io=develop&librarymanagement=develop&util=pull/208/head&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

File: util-cache/src/main/scala/sbt/util/FileInfo.scala [hunk 1]
```diff
@@ -4,13 +4,18 @@
...
-import CacheImplicits._
+import sjsonnew.{ Builder, DeserializationException, JsonFormat, Unbuilder, deserializationError }
+import CacheImplicits.{ arrayFormat => _, _ }
 
 sealed trait FileInfo { def file: File }
-sealed trait HashFileInfo extends FileInfo { def hash: List[Byte] }
+sealed trait HashFileInfo extends FileInfo {
+  @deprecated("Use hashArray instead", "1.3.0")
+  def hash: List[Byte] = hashArray.toList
+  private[util] def hashArray: Array[Byte]
```
Comment (eed3si9n) [inline util-cache/src/main/scala/sbt/util/FileInfo.scala:17]: Should this be public if you want people migrate to hashArray?

**Review by eed3si9n — CHANGES_REQUESTED** [review]
overall looks good.
1 minor question.

File: util-cache/src/main/scala/sbt/util/FileInfo.scala (hunk 1 above)
Comment (jroper) [reply util-cache/src/main/scala/sbt/util/FileInfo.scala:17]: I made it private to ensure immutability. I'm not sure whether these traits should be public or not, I don't see any reason for them to be used directly.

Comment (eed3si9n) [reply util-cache/src/main/scala/sbt/util/FileInfo.scala:17]: `FileInfo.hash` is public, and it returns `HashFileInfo`. I guess there's no reason peope would be called `hash` directly, so it should be ok.

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1724/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: develop-develop-develop-develop-pull/208/head)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1724/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/develop">develop             </A> | sbt/sbt@c124bc1dcdda1c6de55187b75e80194371ca053c
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@4b414b66778399f044beccf749073b52e63bf5ef
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@09658c73b77d72fd5f3ebb71138f033b07f93b12
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@bc2650c55116a5311541d322f16bd2934336d1e3
  util              | <A HREF="https://github.com/sbt/util/pull/208">pull/208/head       </A> | sbt/util@50b2ea6f83073d5ccd1be17dea4833533b8f7620

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=develop&zinc=develop&io=develop&librarymanagement=develop&util=pull/208/head&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Review by eed3si9n — APPROVED** [review]

File: build.sbt [hunk 2]
```diff
@@ -152,6 +152,19 @@ lazy val utilCache = (project in file("util-cache"))
       Seq(sjsonnewScalaJson.value, sjsonnewMurmurhash.value, scalaReflect.value),
     libraryDependencies ++= Seq(scalaTest),
     mimaSettings,
+    mimaBinaryIssueFilters ++= Seq(
+      // These are private case classes that have changed
```
Comment (eed3si9n) [inline build.sbt:156]: FYI - https://github.com/lucidsoftware/neo-sbt-scalafmt/issues/80

**Pull request merged**